nixos/modules/services/hardware/nvidia-container-toolkit.nix

```diff
@@ -0,0 +1,56 @@
+
+  config = lib.mkIf config.hardware.nvidia-container-toolkit.enable {
+    systemd.services.nvidia-container-toolkit = {
+      description = "Container Device Interface (CDI) for Nvidia generator";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "systemd-udev-settle.service" ];
+      serviceConfig = {
+        ExecStart = let
+          script = pkgs.writeScriptBin "nvidia-cdi-generator" ''
+            #! ${pkgs.runtimeShell}
+            mkdir -p /run/cdi
+            ${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk cdi generate \
+              --format json \
+              --ldconfig-path ${pkgs.glibc.bin}/bin/ldconfig \
+              --library-search-path ${pkgs.linuxPackages.nvidia_x11}/lib \
+              --nvidia-ctk-path ${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-cuda-mps-control", "containerPath": "/usr/bin/nvidia-cuda-mps-control", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-cuda-mps-server", "containerPath": "/usr/bin/nvidia-cuda-mps-server", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-debugdump", "containerPath": "/usr/bin/nvidia-debugdump", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-powerd", "containerPath": "/usr/bin/nvidia-powerd", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "/run/current-system/sw/bin/nvidia-smi", "containerPath": "/usr/bin/nvidia-smi", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "${pkgs.glibc}/lib", "containerPath": "${pkgs.glibc}/lib", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "${pkgs.linuxPackages.nvidia_x11}/lib", "containerPath": "${pkgs.linuxPackages.nvidia_x11}/lib", "options": ["ro", "nosuid", "nodev", "bind"] }' | \
+            ${pkgs.jq}/bin/jq \
+               '.containerEdits.mounts[.containerEdits.mounts | length] |= . + { "hostPath": "${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk", "containerPath": "/usr/bin/nvidia-ctk", "options": ["ro", "nosuid", "nodev", "bind"] }' > /run/cdi/nvidia-container-toolkit.json
```

Comment:
Are these the paths that would usually mounted if someone uses the ctk on, let's say, ubuntu? AFAIR the default that the nix version generates is `/run/current-system/sw/bin/xxx` which probably doesn't act nicely with 99% of tools. The `/usr/bin/` seems better.

Reply:
> Are these the paths that would usually mounted if someone uses the ctk on, let's say, ubuntu? 

I haven't checked, but looking at some parts of the source code, `/usr/bin` seems like a [sensible place](https://github.com/NVIDIA/nvidia-container-toolkit/blob/21ed60bc460574d0ed51620375dde270558bec0f/pkg/nvcdi/driver-wsl.go#L114-L117).

> AFAIR the default that the nix version generates is /run/current-system/sw/bin/xxx which probably doesn't act nicely with 99% of tools. The /usr/bin/ seems better.

I don't have a preference for the realisation at `/nix/store` or `/run/current-system/sw/bin` for the tools binaries. However, `nvidia-ctk` is not necessarily in `/run/curent-system/sw/bin`.

Reply:
`pkgs.linuxPackages.nvidia_x11` is more or less always the wrong package to use: we need `config.kernelPackages` instead of `pkgs.linuxPackages`, but also the user might be consuming `nvidia_x11-beta` or something instead, so it's `hardware.nvidia.package` iirc

Reply:
...and neither currently behave nicely on system upgrades, but `addDriverRunpath.driverLink` has better odds of being supported in the future

Reply:
Thank you @SomeoneSerge, I updated it to use `config.hardware.nvidia.package`. I believe this is the right one, thank you for the pointer.